src/controllers/orderController.js

```diff
@@ -5,58 +5,77 @@ const { BadRequestError } = require('../errors');
 const PurchasedAlbum = require('../models/PurchasedAlbum');
 const CustomError = require('../errors');
 const { checkPermissions } = require('../utils');
+const mongoose = require('mongoose');
 
 const createOrder = async (req, res) => {
-  const { orderItems, subtotal, tax, total } = req.body;
+  const session = await mongoose.startSession();
+
+  try {
+    session.startTransaction();
+
+    const { orderItems, subtotal, tax, total } = req.body;
+
+    if (!orderItems?.length) {
+      throw new BadRequestError(
+        'Unable to create an order because the order items are missing.'
+      );
+    }
+    // Fetch user information from the request object
+    const user = req.user;
+
+    const order = new Order({
+      user: req.user.userId,
+      orderItems, // Use the array with full album data
+      subtotal,
+      tax,
+      total,
+    });
 
-  // Perform necessary checks and validations
-  if (!orderItems?.length) {
-    throw new BadRequestError(
-      'Unable to create order because the order items are missing.'
-    );
-  }
+    await order.save({ session });
+
+    for (const orderItem of orderItems) {
+      const purchasedAlbum = new PurchasedAlbum({
+        album: orderItem.album,
+        user: req.user.userId,
+      });
+
+      await purchasedAlbum.save({ session });
+    }
+
+    const paymentIntent = await stripe.paymentIntents.create({
+      amount: total * 100,
+      currency: 'usd',
+      description: `Order #${order._id}`,
+      metadata: {
+        userId: req.user.userId,
+        orderId: order._id,
+        totalQuantity: orderItems.reduce(
+          (total, item) => total + item.quantity,
+          0
+        ),
+        totalAlbums: orderItems.length,
+      },
+    });
 
-  // Create an Order document in the database
-  const order = await Order.create({
-    user: req.user.userId,
-    orderItems,
-    subtotal,
-    tax,
-    total,
-  });
+    order.paymentIntentId = paymentIntent.id;
+    //this is where the payment happens? so if it's successful -we change orderStatus to'complete'? then we should save this new status is a db - that triggers a save middleware in Order.js that sends email of order being paid?
```

Comment:
Not quite; so currently, the flow with Stripe on the backend and front end is as shown in the image in this ticket: [https://codethedreamteam4-2023.atlassian.net/browse/PS-90](https://codethedreamteam4-2023.atlassian.net/browse/PS-90)

**Step 1 (Frontend):**

*   User clicks "Proceed to Checkout" on their cart/order summary. The frontend directs them to the `/checkout` page, and will make a POST request to the `/api/v1/orders` endpoint

**Step 2 (Backend):**

*   Server will _**create**_ an order in the **pending** state. 
*   Server will tell Stripe that an order is being made (creating a "paymentIntent" object in Stripe)
*   Stripe will give the server the `paymentIntent` object. The server will _**save**_ the `id` of this object in the Order document so that we can easily associate the two. 
*   The server returns the order object, and the `clientSecret` from Stripe's paymentIntent object as part of the response

**Step 3 (Frontend):**

*   Frontend will pass the `clientSecret` to Stripe. This is how Stripe authenticates and knows that this frontend is legitimately handling a Stripe payment for the user
*   Frontend shows payment page to the user
*   User inputs their payment info (eg: credit card)
*   User submits their payment
    *   _**Note:**_ If the user doesn't submit their payment (they forgot or leave the tab for too long; eventually our `setInterval` in the backend will mark the order as 'cancelled' since they've abandoned it for too long)
*   Frontend sends the user payment info to Stripe (along with the `clientSecret` and a `redirectUrl`)
*   After Stripe processes the payment, it will cause the browser to redirect to the checkout complete page (which we gave as the `redirectUrl`)

**Step 4 (Frontend):**

*   Frontend will call `GET /api/v1/orders/:id` to get order details and will ask Stripe for the paymentIntent so that it can tell the user if the payment was successful or not. 

So for your question of "how does the backend know if the order payment is successful" we have at least two options:

**Option 1: Frontend tells the server**

*   At **Step 4** the frontend has info from Stripe about whether the purchase was successful or not. At this point it can make a `PATCH /api/v1/orders/:id` call to update the orderStatus to `payment_successful`
    *   We don't currently have this endpoint implemented in the backend
    *   We don't currently have a ticket to do this work in the frontend

**Option 2: Backend listens to a webhook from Stripe**

*   This would be this ticket: https://codethedreamteam4-2023.atlassian.net/browse/PS-202

Reply:
@akosasante Akos, thank you for such a great explanation!!! finally undestood it and assigned to myself PS 202 to see how his stripe hook works

Reply:
https://stripe.com/docs/api/payment_intents/object - stripe docs

Reply:
:+1:  remember to remove this line 61 comment to avoid future confusion